fix(merge-on-green): add cleanup cron job, fix blocking issue for removing

sofisl opened the pull request:
See [java-bigtable/498](https://github.com/googleapis/java-bigtable/pull/498), and [this issue](https://github.com/googleapis/repo-automation-bots/issues/1088).

I did some further digging into this multiple commenting issue, that did not seem to be solved by #1091. I think the real issue here is when people add an 'automerge' label and then either remove that label, close, or merge that PR soon after. I think a race condition is happening where the PR is not yet recorded in the Datastore table when the PR is closed or removed, making so that it doesn't actually get removed. 

I don't know how to solve this problem categorically. What I ended up doing is creating a clean-up cron job that runs every 30 minutes. Every 30 minutes, if the cron job is the 'cleanup' job, the app will go through the PRs in the Datastore table and clean up ones that are either closed, merged, or do not have an automerge label on them anymore.

tbpg commented on packages/merge-on-green/src/merge-on-green.ts:
```diff
@@ -148,6 +148,59 @@ handler.cleanUpPullRequest = async function cleanUpPullRequest(
 };
 
+/**
+ * Check if PR has been merged, closed, or unlabeled, then remove
```
Missing end of sentence?

tbpg reviewed (changes requested):
Design question: could we do this cleanup "check" before doing anything with a PR in the normal loop/cron? That way, there won't be a delay between the problematic state and when it gets fixed by the next cleanup cron iteration.

Also, I saw the multiple-comment behavior on https://github.com/googleapis/repo-automation-bots/pull/1111, which doesn't seem like the same issue?

sofisl commented:
Thanks for the review @tbpg!
> Design question: could we do this cleanup "check" before doing anything with a PR in the normal loop/cron? That way, there won't be a delay between the problematic state and when it gets fixed by the next cleanup cron iteration.

I think what you're asking is if we can check to see if the PR is closed/merged/unlabeled before doing something in the regular workflow. This was the previous implementation of the bot, but I changed it in #1023 because checking this info was causing us to run into rate limits very quickly. Now, we wait for the unlabeled/closed/merged event to update the Datastore table. But, now we're running into a separate issue on whether or not the PR is in the table before it gets deleted if it gets added and deleted in the table too quickly.

But, maybe we run the cleanup cron job every 15 minutes, instead of every 30? It is an expensive task to do, but it might be worth to address these issues. I'll leave that suggestion up to you and/or anyone else who has thoughts on it :)

> Also, I saw the multiple-comment behavior on #1111, which doesn't seem like the same issue?

Funny you mention that PR, it is actually the same issue! I was testing it last night. I quickly added and removed the automerge label. 

![Screen Shot 2020-11-12 at 12 19 47 PM](https://user-images.githubusercontent.com/55454395/98991999-75f27e80-24e1-11eb-99b4-fa2b9312f9d0.png)

If you see there, MOG actually ended up merging that PR, even though it didn't have an automerge label on it. I am not exactly clear on what causes the multiple commenting issue, but I do know that this fix should solve it, since that PR shouldn't have even been in the Datastore table to begin with.

tmatsuo commented on packages/sync-repo-settings/src/sync-repo-settings.ts:
```diff
@@ -26,7 +26,7 @@ import Ajv from 'ajv';
 import {PullsListFilesResponseData} from '@octokit/types';
 
-export const configFileName = 'sync-repo-settings.yaml';
+export const configFileName = 'sync-repo-settings.yml';
```
There seem to be some changes unrelated?

sofisl replied:
oof, thanks for the heads up. Need to resync to main branch.

tmatsuo commented on packages/sync-repo-settings/src/sync-repo-settings.ts:
```diff
@@ -340,12 +340,6 @@ async function updateRepoTeams(
   logger.info(`Update team access for ${repo}`);
   const [owner, name] = repo.split('/');
 
```
Is this also an irrelevant change?

bcoe commented on packages/merge-on-green/test/merge-on-green.ts:
```diff
@@ -798,6 +810,93 @@ describe('merge-on-green', () => {
+        assert(removePRStub.called);
+      });
+
+      it('deletes a PR if label is not found when cleaning up repository', async () => {
```
awesome to see the increased coverage 👍

bcoe commented on packages/merge-on-green/src/merge-on-green.ts:
```diff
@@ -210,15 +297,19 @@ function handler(app: Application) {
             );
             if (remove || wp.state === 'stop') {
+              await handler.removePR(wp.url);
+              try {
```
spacing looks a little weird here, might need to run `npm run fix`.

bcoe commented on packages/merge-on-green/src/merge-on-green.ts:
```diff
@@ -188,6 +252,29 @@ function handler(app: Application) {
   app.on('schedule.repository' as any, async context => {
     const watchedPRs = await handler.listPRs();
+    if (context.payload.cleanUp === true) {
+      while (watchedPRs.length) {
```
I might be tempted to pull this into a helper function.

bcoe reviewed (changes requested):
The logic in `schedule.repository` is getting pretty expansive, perhaps we could split it into two helpers that get called depending on the type of payload:

`runCleanUp`/`runMergeOnGreen` or something like that?

tmatsuo commented:
This is kind of meta comment. Take this a grain of solt.

The PR title should be something like:
```
fix(merge-on-green): add cleanup cron job
```
1. Maybe we should use package name in the paren
2. More concrete title is better in my opinion

tbpg reviewed (approved):
A few small comments. Otherwise, LGTM.

sofisl commented on packages/merge-on-green/src/merge-on-green.ts:
```diff
@@ -135,17 +135,82 @@ handler.cleanUpPullRequest = async function cleanUpPullRequest(
+      })
+    ).data;
+  } catch (err) {
+    pr = undefined;
```
I might leave this just in case the github API is unavailable for some reason, but we can confirm by checking to see if a label is on the PR in the next function.
